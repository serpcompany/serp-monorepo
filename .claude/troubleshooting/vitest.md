https://vitest.dev/guide/common-errors.html