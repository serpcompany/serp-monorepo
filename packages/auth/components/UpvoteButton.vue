<template>
  <div>
    <UButton :color="buttonColor" variant="ghost" :loading="loading" icon="i-lucide-arrow-up" @click="upvote" />
    <UBadge>{{ localUpvotes.length }}</UBadge>
  </div>
</template>

<script setup lang="ts">
const { status, data } = useAuth();
const props = defineProps<{
  id: number;
  module: string;
  upvotes: string[];
}>();

const localUpvotes = ref<string[]>([]);
const toast = useToast();
const loading = ref(false);

const buttonColor = computed(() =>
  localUpvotes.value.includes(data?.value?.user?.email) ? 'primary' : 'neutral'
);

watch(
  () => props.upvotes,
  (newUpvotes) => {
    localUpvotes.value = [...newUpvotes];
  },
  { immediate: true }
);

async function upvote() {
  try {
    loading.value = true;

    if (status.value === 'authenticated') {
      if (!data?.value?.user?.email) {
        throw new Error('User not authenticated');
      }

      const { data: response, error } = await useFetch('/api/upvote', {
        method: 'POST',
        headers: useRequestHeaders(['cookie']),
        body: JSON.stringify({ id: props.id, module: props.module })
      });

      if (error.value) {
        throw new Error(`Failed to upvote - ${error.value.message}`);
      }

      if (response.value.message === 'success') {
        if (localUpvotes.value.includes(data.value.user.email)) {
          localUpvotes.value = localUpvotes.value.filter(
            (email) => email !== data.value.user.email
          );
        } else {
          localUpvotes.value.push(data.value.user.email);
        }

        toast.add({
          id: 'upvote-success',
          title: 'Upvoted',
          description: 'Your upvote has been recorded',
          icon: 'check-circle'
        });
      } else {
        toast.add({
          id: 'upvote-error',
          title: 'Error upvoting',
          description: response.value.message,
          icon: 'exclamation-circle'
        });
      }
    } else {
      toast.add({
        id: 'upvote-login',
        title: 'Login required',
        description: 'You need to login to upvote',
        icon: 'exclamation-circle'
      });
    }
  } catch (error) {
    toast.add({
      id: 'upvote-error',
      title: 'Error upvoting',
      description: error.message,
      icon: 'exclamation-circle'
    });
  } finally {
    loading.value = false;
  }
}
</script>
