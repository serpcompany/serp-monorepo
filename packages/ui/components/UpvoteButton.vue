<template>
  <div class="flex w-full items-center sm:w-auto">
    <button
      class="flex w-full items-center justify-center gap-2 rounded-full border border-gray-300 px-4 py-2 font-medium transition-colors hover:bg-gray-100 disabled:opacity-75 sm:w-auto dark:border-gray-600 dark:hover:bg-gray-800"
      :class="{
        'text-orange-500 dark:text-orange-400': localUpvotes.includes(
          data?.value?.user?.email
        )
      }"
      :disabled="loading"
      @click="upvote"
    >
      <span
        v-if="loading"
        class="inline-block h-4 w-4 animate-spin rounded-full border-2 border-solid border-current border-r-transparent"
      ></span>
      <svg
        v-else
        class="h-4 w-4"
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 24 24"
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        stroke-linecap="round"
        stroke-linejoin="round"
      >
        <path d="m5 12 7-7 7 7" />
        <path d="M12 19V5" />
      </svg>
      <span>Upvotes {{ localUpvotes.length }}</span>
    </button>
  </div>
</template>

<script setup lang="ts">
const { status, data } = useAuth();
const props = defineProps<{
  id: number;
  module: string;
  upvotes: string[];
}>();

const localUpvotes = ref<string[]>([]);
const toast = useToast();
const loading = ref(false);

const buttonColor = computed(() =>
  localUpvotes.value.includes(data?.value?.user?.email) ? 'primary' : 'neutral'
);

watch(
  () => props.upvotes,
  (newUpvotes) => {
    if (newUpvotes && newUpvotes.length) {
      localUpvotes.value = [...newUpvotes];
    } else {
      localUpvotes.value = [];
    }
  },
  { immediate: true }
);

async function upvote() {
  try {
    loading.value = true;

    if (status.value === 'authenticated') {
      if (!data?.value?.user?.email) {
        throw new Error('User not authenticated');
      }

      const { data: response, error } = await useFetch('/api/upvote', {
        method: 'POST',
        headers: useRequestHeaders(['cookie']),
        body: JSON.stringify({ id: props.id, module: props.module })
      });

      if (error.value) {
        throw new Error(`Failed to upvote - ${error.value.message}`);
      }

      if (response.value.message === 'success') {
        if (localUpvotes.value.includes(data.value.user.email)) {
          localUpvotes.value = localUpvotes.value.filter(
            (email) => email !== data.value.user.email
          );
        } else {
          localUpvotes.value.push(data.value.user.email);
        }

        toast.add({
          id: 'upvote-success',
          title: 'Upvoted',
          description: 'Your upvote has been recorded',
          icon: 'check-circle'
        });
      } else {
        toast.add({
          id: 'upvote-error',
          title: 'Error upvoting',
          description: response.value.message,
          icon: 'exclamation-circle'
        });
      }
    } else {
      toast.add({
        id: 'upvote-login',
        title: 'Login required',
        description: 'You need to login to upvote',
        icon: 'exclamation-circle'
      });
    }
  } catch (error) {
    toast.add({
      id: 'upvote-error',
      title: 'Error upvoting',
      description: error.message,
      icon: 'exclamation-circle'
    });
  } finally {
    loading.value = false;
  }
}
</script>
