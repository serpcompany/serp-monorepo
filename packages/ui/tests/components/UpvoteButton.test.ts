import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { describe, expect, it, vi } from 'vitest';
import UpvoteButton from '../../components/UpvoteButton.vue';
import ComponentRender from '../componentRender';
import '../mockUseUserSession';

mockNuxtImport('useFetch', () => () => ({
  data: ref({ message: 'success' }),
  error: ref(null)
}));

mockNuxtImport('useToast', () => () => ({
  add: vi.fn()
}));

mockNuxtImport('useRequestHeaders', () => () => ({}));

describe('UpvoteButton', () => {
  const baseProps = { id: 1, module: 'posts' };

  it.each([
    // Testing with an empty upvotes array
    ['with no upvotes', { props: { ...baseProps, upvotes: [] } }],
    // Testing with an upvotes array that does not contain the user's email
    [
      'with upvotes not containing user email',
      { props: { ...baseProps, upvotes: ['other@example.com'] } }
    ],
    // Testing with an upvotes array that includes the user's email
    [
      'with upvotes containing user email',
      { props: { ...baseProps, upvotes: ['test@test.com'] } }
    ]
  ])(
    'renders %s correctly',
    async (desc: string, options: { props: unknown }) => {
      const html = await ComponentRender(
        `UpvoteButton ${desc}`,
        options,
        UpvoteButton
      );
      expect(html).toMatchSnapshot();
    }
  );
});
