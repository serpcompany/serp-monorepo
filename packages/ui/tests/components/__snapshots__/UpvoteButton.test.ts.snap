// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpvoteButton > renders with no upvotes correctly 1`] = `
<div class="
  flex
  items-center
  sm:w-auto
  w-full
">
  <button
    class="
      aria-disabled:bg-transparent
      aria-disabled:cursor-not-allowed
      aria-disabled:opacity-75
      dark:aria-disabled:bg-transparent
      dark:disabled:bg-transparent
      disabled:bg-transparent
      disabled:cursor-not-allowed
      disabled:opacity-75
      focus-visible:ring-2
      focus-visible:ring-primary
      focus:outline-none
      font-medium
      gap-1.5
      hover:bg-primary/10
      inline-flex
      items-center
      px-2.5
      py-1.5
      ring
      ring-inset
      ring-primary/50
      rounded-md
      text-primary
      text-sm
      transition-colors
    "
    type="button"
  >
    <!-- v-if -->
    <span class="
      flex
      items-center
      text-xl
    ">
      <span
        aria-hidden="true"
        class="
          i-lucide:square-chevron-up
          iconify
          mr-2
          size-5
        "
      ></span>
      <span class="text-lg">
        0
      </span>
    </span>
    <!-- v-if -->
  </button>
</div>
`;

exports[`UpvoteButton > renders with upvotes containing user email correctly 1`] = `
<div class="
  flex
  items-center
  sm:w-auto
  w-full
">
  <button
    class="
      aria-disabled:bg-transparent
      aria-disabled:cursor-not-allowed
      aria-disabled:opacity-75
      dark:aria-disabled:bg-transparent
      dark:disabled:bg-transparent
      disabled:bg-transparent
      disabled:cursor-not-allowed
      disabled:opacity-75
      focus-visible:ring-2
      focus-visible:ring-primary
      focus:outline-none
      font-medium
      gap-1.5
      hover:bg-primary/10
      inline-flex
      items-center
      px-2.5
      py-1.5
      ring
      ring-inset
      ring-primary/50
      rounded-md
      text-primary
      text-sm
      transition-colors
    "
    type="button"
  >
    <!-- v-if -->
    <span class="
      flex
      items-center
      text-xl
    ">
      <span
        aria-hidden="true"
        class="
          i-lucide:square-chevron-up
          iconify
          mr-2
          size-5
        "
      ></span>
      <span class="text-lg">
        1
      </span>
    </span>
    <!-- v-if -->
  </button>
</div>
`;

exports[`UpvoteButton > renders with upvotes not containing user email correctly 1`] = `
<div class="
  flex
  items-center
  sm:w-auto
  w-full
">
  <button
    class="
      aria-disabled:bg-transparent
      aria-disabled:cursor-not-allowed
      aria-disabled:opacity-75
      dark:aria-disabled:bg-transparent
      dark:disabled:bg-transparent
      disabled:bg-transparent
      disabled:cursor-not-allowed
      disabled:opacity-75
      focus-visible:ring-2
      focus-visible:ring-primary
      focus:outline-none
      font-medium
      gap-1.5
      hover:bg-primary/10
      inline-flex
      items-center
      px-2.5
      py-1.5
      ring
      ring-inset
      ring-primary/50
      rounded-md
      text-primary
      text-sm
      transition-colors
    "
    type="button"
  >
    <!-- v-if -->
    <span class="
      flex
      items-center
      text-xl
    ">
      <span
        aria-hidden="true"
        class="
          i-lucide:square-chevron-up
          iconify
          mr-2
          size-5
        "
      ></span>
      <span class="text-lg">
        1
      </span>
    </span>
    <!-- v-if -->
  </button>
</div>
`;
